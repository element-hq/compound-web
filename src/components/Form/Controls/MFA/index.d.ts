export { MFAControl } from "./MFA";
//# sourceMappingURL=index.d.ts.map