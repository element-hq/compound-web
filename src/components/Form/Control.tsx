/*
Copyright 2023  New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { PropsWithChildren } from "react";
import { Control as RadixControl } from "@radix-ui/react-form";

import styles from "./form.module.css";
import classNames from "classnames";

type ControlProps = {
  className?: string;
} & React.ComponentProps<typeof RadixControl>;

/**
 * Thin wrapper around Radix UI Control component
 * https://www.radix-ui.com/docs/primitives/components/form#control
 */
export function Control({
  children,
  ...props
}: PropsWithChildren<ControlProps>): JSX.Element {
  const classes = classNames(styles.control, props.className);
  return (
    <RadixControl {...props} className={classes}>
      {children}
    </RadixControl>
  );
}
