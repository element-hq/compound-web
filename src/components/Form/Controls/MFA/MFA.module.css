/*
Copyright 2023  New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.mfa {
  inline-size: fit-content;
  display: flex;
  flex-direction: row;
  gap: 1rem;
  position: relative;
}

.mfa-input {
  opacity: 0;
  position: absolute;
  inset: 0;
  inline-size: 100%;
  block-size: 100%;
  box-sizing: border-box;
}

.digit {
  inline-size: 3rem;
  aspect-ratio: 1 / 1;
  border: 1px solid grey;
  text-align: center;
}
