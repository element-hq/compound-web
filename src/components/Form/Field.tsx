/*
Copyright 2023 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React, { forwardRef, type PropsWithChildren } from "react";
import { Field as RadixField } from "@radix-ui/react-form";

import styles from "./form.module.css";
import classNames from "classnames";

type FieldProps = {
  /**
   * The CSS class name.
   */
  className?: string;
} & React.ComponentProps<typeof RadixField>;

/**
 * Thin wrapper around Radix UI Field component
 * https://www.radix-ui.com/docs/primitives/components/form#field
 */
export const Field = forwardRef<HTMLDivElement, PropsWithChildren<FieldProps>>(
  function Field({ children, ...props }, ref) {
    const classes = classNames(styles.field, props.className);
    return (
      <RadixField ref={ref} {...props} className={classes}>
        {children}
      </RadixField>
    );
  },
);
