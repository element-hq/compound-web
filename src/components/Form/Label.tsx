/*
Copyright 2023  New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { PropsWithChildren } from "react";
import { Label as RadixLabel } from "@radix-ui/react-form";

import styles from "./form.module.css";
import classNames from "classnames";

type LabelProps = {
  className?: string;
} & React.ComponentProps<typeof RadixLabel>;

/**
 * Thin wrapper around Radix UI Label component
 * https://www.radix-ui.com/docs/primitives/components/form#label
 */
export function Label({
  children,
  ...props
}: PropsWithChildren<LabelProps>): JSX.Element {
  const classes = classNames(styles.label, props.className);
  return (
    <RadixLabel {...props} className={classes}>
      {children}
    </RadixLabel>
  );
}
