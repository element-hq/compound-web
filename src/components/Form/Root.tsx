/*
Copyright 2023 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React, { forwardRef, PropsWithChildren } from "react";
import { Root as RadixRoot } from "@radix-ui/react-form";

import styles from "./form.module.css";
import classNames from "classnames";

type RootProps = {
  /**
   * The CSS class name.
   */
  className?: string;
} & React.ComponentProps<typeof RadixRoot>;

/**
 * Thin wrapper around Radix UI Root component
 * https://www.radix-ui.com/docs/primitives/components/form#root
 */
export const Root = forwardRef<HTMLFormElement, PropsWithChildren<RootProps>>(
  function Root({ children, ...props }, ref) {
    const classes = classNames(styles.root, props.className);
    return (
      <RadixRoot ref={ref} {...props} className={classes}>
        {children}
      </RadixRoot>
    );
  },
);
