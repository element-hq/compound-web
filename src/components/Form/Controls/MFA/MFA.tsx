/*
Copyright 2023  New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { PropsWithoutRef, useState } from "react";
import classnames from "classnames";
import { Control } from "../../Control";

import styles from "./MFA.module.css";
import formStyles from "../../form.module.css";
import classNames from "classnames";

type MFAProps = {
  className?: string;
  length?: number;
  value?: string;
  disabled?: boolean;
} & React.ComponentProps<typeof Control>;

/**
 * Thin wrapper around Radix UI Control component
 * https://www.radix-ui.com/docs/primitives/components/form#control
 */
export const MFAControl = React.forwardRef(function MFAControl(
  { className, length = 6, ...props }: PropsWithoutRef<MFAProps>,
  ref: React.ForwardedRef<HTMLInputElement>,
) {
  const classes = classNames(styles.mfa, className);

  const [mfaDigits, setMfaDigits] = useState(props.value ?? "");

  return (
    <div className={classes}>
      <input
        {...props}
        // Showing digits on mobile browsers. Using numbers is not really suited
        // as it often adds a way to increment or decrement the current value
        // which is not interesting for this use case
        type="tel"
        minLength={0}
        maxLength={length}
        value={mfaDigits}
        className={styles["mfa-input"]}
        onChange={(e) => {
          setMfaDigits(
            e.currentTarget.value.replace(/[^\d]/g, "").slice(0, length),
          );
        }}
        autoComplete="off"
        ref={ref}
      />
      {Array(length)
        .fill(0)
        .map((_, index) => {
          return (
            <div
              key={`digit-${index}`}
              className={classnames(styles.digit, formStyles.control)}
              aria-hidden="true"
            >
              {mfaDigits.charAt(index) ?? ""}
            </div>
          );
        })}
    </div>
  );
});
