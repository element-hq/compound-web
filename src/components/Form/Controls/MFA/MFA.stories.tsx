/*
Copyright 2025 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.
Copyright 2023 New Vector Ltd

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React from "react";

import { Meta, StoryObj } from "@storybook/react";

import { MFAInput } from "./MFA";

type Props = React.ComponentProps<typeof MFAInput> & {
  invalid?: boolean;
};

const meta = {
  title: "Form/Controls/MFA",
  component: MFAInput,
  render: ({ invalid, ...props }) => (
    <MFAInput {...props} data-invalid={invalid || undefined} />
  ),
  tags: ["autodocs"],
  parameters: {
    controls: {
      include: ["defaultValue", "disabled", "readOnly", "invalid"],
    },
  },
  argTypes: {
    defaultValue: {
      type: "string",
    },
    autoFocus: {
      type: "boolean",
    },
    disabled: {
      type: "boolean",
    },
    readOnly: {
      type: "boolean",
    },
  },
  args: {
    defaultValue: "",
    autoFocus: false,
    disabled: false,
    readOnly: false,
    invalid: false,
  },
} satisfies Meta<Props>;
export default meta;

type Story = StoryObj<typeof meta>;

export const Empty: Story = {
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=793-4786",
    },
  },
};

export const Filled: Story = {
  args: { defaultValue: "012345" },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=793-4827",
    },
  },
};

export const Disabled: Story = {
  args: {
    defaultValue: "012345",
    disabled: true,
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=793-4821",
    },
  },
};

export const ReadOnly: Story = {
  args: {
    defaultValue: "012345",
    readOnly: true,
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=2126-21991",
    },
  },
};

export const Focus: Story = {
  args: {
    defaultValue: "012",
    autoFocus: true,
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=793-4805",
    },
  },
};

export const Invalid: Story = {
  args: {
    invalid: true,
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=793-4814",
    },
  },
};
