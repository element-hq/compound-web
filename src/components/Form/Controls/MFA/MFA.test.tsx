/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { act, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { MFAControl } from "./MFA";
import { Root } from "../../Root";
import { Field } from "../../Field";
import { Label } from "../../Label";
import { Control } from "../../Control";

describe("PasswordControl", () => {
  function getMfa(
    props: React.ComponentProps<typeof MFAControl> = {}
  ): JSX.Element {
    return (
      <Root>
        <Field name="mfa">
          <Label>MFA</Label>
          <Control asChild>
            <MFAControl {...props} placeholder="000000" />
          </Control>
        </Field>
      </Root>
    );
  }

  it("renders", () => {
    const { asFragment } = render(getMfa());
    expect(asFragment()).toMatchSnapshot();
  });

  it("only allows digits", async () => {
    const { getByLabelText } = render(getMfa());

    const input = getByLabelText("MFA");
    await act(async () => {
      await userEvent.type(input, "12sf65");
    });

    expect(input).toHaveValue("1265");
  });

  it("respects max length", async () => {
    const { getByLabelText } = render(getMfa({ length: 3 }));

    const input = getByLabelText("MFA");
    await act(async () => {
      await userEvent.type(input, "67890");
    });

    expect(input).toHaveValue("678");
  });
});
