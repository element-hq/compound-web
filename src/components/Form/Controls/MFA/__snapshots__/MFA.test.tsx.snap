// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PasswordControl > renders 1`] = `
<DocumentFragment>
  <form
    class="_root_dc87d2"
  >
    <div
      class="_field_dc87d2"
    >
      <label
        class="_label_dc87d2"
        for="radix-:r0:"
      >
        MFA
      </label>
      <div
        class="_mfa_4b8c2b _control_dc87d2"
      >
        <input
          autocomplete="off"
          class="_mfa-input_4b8c2b"
          id="radix-:r0:"
          maxlength="6"
          minlength="0"
          name="mfa"
          placeholder="000000"
          title=""
          type="tel"
          value=""
        />
        <div
          aria-hidden="true"
          class="_digit_4b8c2b _control_dc87d2"
        />
        <div
          aria-hidden="true"
          class="_digit_4b8c2b _control_dc87d2"
        />
        <div
          aria-hidden="true"
          class="_digit_4b8c2b _control_dc87d2"
        />
        <div
          aria-hidden="true"
          class="_digit_4b8c2b _control_dc87d2"
        />
        <div
          aria-hidden="true"
          class="_digit_4b8c2b _control_dc87d2"
        />
        <div
          aria-hidden="true"
          class="_digit_4b8c2b _control_dc87d2"
        />
      </div>
    </div>
  </form>
</DocumentFragment>
`;
